### PR TITLE
Add unit testing

### PR DESCRIPTION
**Pull request by trevorhuis-smoothstack**
Hey I am sorry for the big PR, i thought i had one up for some of this. I refactored my application a bit to break the services and tests into smaller files. Almost all of the new code is unit testing and some changes that I found were needed while writing the tests.

### CODE REVIEW

File: src/main/java/com/ss/training/utopia/agent/service/AgentReadService.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package com.ss.training.utopia.agent.service;
```
Comment (trevorhuis-smoothstack) [inline src/main/java/com/ss/training/utopia/agent/service/AgentReadService.java:1]: No new code, just seperate file

File: src/main/java/com/ss/training/utopia/agent/service/AgentCancelService.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package com.ss.training.utopia.agent.service;
```
Comment (trevorhuis-smoothstack) [inline src/main/java/com/ss/training/utopia/agent/service/AgentCancelService.java:1]: No new code, just seperate file

File: src/main/java/com/ss/training/utopia/agent/service/AgentBookingService.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
+package com.ss.training.utopia.agent.service;
```
Comment (trevorhuis-smoothstack) [inline src/main/java/com/ss/training/utopia/agent/service/AgentBookingService.java:1]: No new code, just seperate file

File: src/main/java/com/ss/training/utopia/agent/controller/AgentController.java [hunk 4]
```diff
@@ -24,20 +29,33 @@
...
+	@Autowired UserService userService;
 
 
     @GetMapping(path="/airports")
     public ResponseEntity<Airport[]> getAllAirports() {
 		List<Airport> airportList = null;
 		Airport[] airportArray = null;
+		HttpStatus status = HttpStatus.OK;
+		airportList = readService.readAirports();
+		if (airportList == null) // no airports exist in the database
```
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/utopia/agent/controller/AgentController.java]: This actually would mean the DAO threw an exception. If no airports are in the database the service will return a list of size 0.

File: src/main/java/com/ss/training/utopia/agent/service/UserService.java [hunk 5]
```diff
@@ -2,46 +2,32 @@
...
-		User user;
-		try {
-			user = userDao.findByUsername(username);
-		} catch (Throwable t) {
-			return null;
-		}
-		return (user != null && "TRAVELER".equals(user.getRole()));
+	public User getUser(String username) {
+		User user = userDao.findByUsername(username);
+		user.setPassword(null);
```
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/utopia/agent/service/UserService.java:32]: 👍

File: src/test/java/com/ss/training/utopia/agent/service/UserServiceTests.java [hunk 6]
```diff
@@ -41,23 +40,8 @@ public void createUserTest() {
 		String password = "password";
 		User user = new User(null, null, null, password, null);
 		Mockito.when(userDAO.save(user)).thenReturn(null);
-		assertTrue(userService.createUser(user));
+		assertEquals(userService.createUser(user), user);
 		assertTrue(new BCryptPasswordEncoder().matches(password, user.getPassword()));
 		Mockito.when(userDAO.save(user)).thenThrow(new RuntimeException());
```
Comment (justin-obrien-smoothstack) [inline src/test/java/com/ss/training/utopia/agent/service/UserServiceTests.java:45]: This line is no longer serving any purpose.

File: src/main/java/com/ss/training/utopia/agent/service/UserService.java [hunk 7]
```diff
@@ -2,46 +2,32 @@
...
 		user.setPassword(new BCryptPasswordEncoder().encode(user.getPassword()));
-		try {
-			userDao.save(user);
-		} catch (Throwable t) {
-			return null;
-		}
-		return true;
-	}
+		userDao.save(user);
+		return user;
```
Comment (justin-obrien-smoothstack) [inline src/main/java/com/ss/training/utopia/agent/service/UserService.java]: I don't think the returned user will include the user ID. Not sure though.

File: src/test/java/com/ss/training/utopia/agent/DAOTestSuite.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+		
+		Flight flightOne = new Flight(oneLong, twoLong, timestamp, oneLong, (short) 0, null);
+		Flight flightTwo = new Flight(twoLong, oneLong, timestamp, twoLong, (short) 0, null);
+        
+        flightDAO.save(flightOne);
+        flightDAO.save(flightTwo);     
+
+		assertEquals(flightOne, flightDAO.findByFlightId(oneLong));
+		assertNull(flightDAO.findByFlightId(threeLong));
+	}
```
Comment (justin-obrien-smoothstack) [inline src/test/java/com/ss/training/utopia/agent/DAOTestSuite.java:47]: This is a duplicate of the method in FlightDAOTestSuite.

File: src/test/java/com/ss/training/utopia/agent/dao/FlightDAOTestSuite.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+       flightDAO.save(otherFutureFlight);
+       flightDAO.save(otherFutureFlightNoSeats);
+
+       List<Flight> availableFlights = flightDAO.findAvailable();
+
+       assertEquals(availableFlights.size(), 2);
+        assertEquals( availableFlights.get(0).getDepartId(), twoLong);
+        assertEquals( availableFlights.get(1).getDepartId(), twoLong);
+        assertNotEquals((short) availableFlights.get(0).getSeatsAvailable(), (short) 0);
+        assertNotEquals((short) availableFlights.get(1).getSeatsAvailable(), (short) 0);
```
Comment (justin-obrien-smoothstack) [inline src/test/java/com/ss/training/utopia/agent/dao/FlightDAOTestSuite.java:84]: The available seats are already shorts; they shouldn't need to be cast.

File: src/test/java/com/ss/training/utopia/agent/service/AgentServiceTestSuite.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+import org.mockito.Spy;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.ActiveProfiles;
+
+
+@SpringBootTest(classes = { AgentApplication.class, H2TestProfileJPAConfig.class })
+@ActiveProfiles("test")
+@RunWith(MockitoJUnitRunner.class)
+public class AgentServiceTestSuite {
```
Comment (justin-obrien-smoothstack) [inline src/test/java/com/ss/training/utopia/agent/service/AgentServiceTestSuite.java:35]: I'm guessing you meant to name this AgentReadServiceTestSuite.

File: src/test/java/com/ss/training/utopia/agent/service/BookingServiceTestSuite.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        assertEquals(createBookingResult, "Flight Booked");
+    }
+
+    @Test
+    public void createBookingTransactionTestNoSeats() throws AuthenticationException, InvalidRequestException, APIConnectionException,
+            CardException, APIException {
+
+
+        Booking booking = new Booking(1l, 1l, 1l, true, null);
+        Flight emptyFlight = new Flight(1l, 1l, null, 1l, (short) 0, 10f);
```
Comment (justin-obrien-smoothstack) [inline src/test/java/com/ss/training/utopia/agent/service/BookingServiceTestSuite.java]: fullFlight seems lime a more fitting name for this variable.

**Review by justin-obrien-smoothstack — CHANGES_REQUESTED** [review]
Looks good.

Changes requested:
- In controller, use empty list as condition for no airports in database, and null list as condition for DAO error

**Pull request merged**